src/Compilers/CSharp/Portable/Lowering/Instrumentation/DynamicAnalysisInjector.cs

```diff
@@ -129,6 +134,79 @@ private static bool IsExcludedFromCodeCoverage(MethodSymbol method)
             return false;
         }
 
+        private static BoundExpressionStatement GetCreatePayloadStatement(
+            ImmutableArray<SourceSpan> dynamicAnalysisSpans,
+            BoundStatement methodBody,
+            LocalSymbol methodPayload,
+            MethodSymbol createPayloadForMethodsSpanningSingleFile,
+            MethodSymbol createPayloadForMethodsSpanningMultipleFiles,
+            BoundExpression mvid,
+            BoundExpression methodToken,
+            BoundExpression payloadSlot,
+            SyntheticBoundNodeFactory methodBodyFactory,
+            DebugDocumentProvider debugDocumentProvider)
+        {
+            MethodSymbol createPayloadOverload;
+            BoundExpression fileIndexOrIndicesArgument;
+
+            if (dynamicAnalysisSpans.IsEmpty)
+            {
+                createPayloadOverload = createPayloadForMethodsSpanningSingleFile;
+
+                // For a compiler generated method that has no 'real' spans, we emit the index for
+                // the document corresponding to the syntax node that is associated with its bound node.
+                fileIndexOrIndicesArgument = methodBodyFactory.SourceDocumentIndex(GetSourceDocument(debugDocumentProvider, methodBody.Syntax));
+            }
+            else
+            {
+                var documents = PooledHashSet<DebugSourceDocument>.GetInstance();
+                foreach (var span in dynamicAnalysisSpans)
+                {
+                    documents.Add(span.Document);
+                }
+
+                // At this point, we should have at least one document since we have already
+                // handled the case where method has no 'real' spans (and therefore no documents) above.
+                if (documents.Count == 1)
+                {
+                    createPayloadOverload = createPayloadForMethodsSpanningSingleFile;
+                    fileIndexOrIndicesArgument = methodBodyFactory.SourceDocumentIndex(documents.Single());
+                }
+                else
+                {
+                    // Sort the documents so that the emit order for the indices is always deterministic.
+                    var sortedDocuments = ArrayBuilder<DebugSourceDocument>.GetInstance(documents.Count);
+                    sortedDocuments.AddRange(documents);
+                    sortedDocuments.Sort(DebugSourceDocumentLocationComparer.Instance);
```

Comment:
This is rendered unnecessary if we do what you proposed above (i.e. rely on dynamicAnalysisSpans already having deterministic order).